### PR TITLE
Add an AA range (debug mode only)

### PR DESCRIPTION
**Pull request by marler8997**
Made this input range for dmd associative arrays for debug purposes.  Decided to make a PR for it if other people think it's useful to include.

### CODE REVIEW

File: src/dmd/root/aav.d [hunk 1]
```diff
@@ -121,6 +121,61 @@ extern (C++) Value dmd_aaGetRvalue(AA* aa, Key key)
...
+        Value value;
+    }
+    /**
+    Returns an input range of key/values from `aa`.
+    Returns: an input range of key/values from `aa`.
+    */
+    auto aaRange(AA* aa)
+    {
+        return AARange(aa);
+    }
```
Comment (jacob-carlborg) [inline src/dmd/root/aav.d]: Please add an empty newline between functions.

**Comment by JinShil** [comment]
What's the problem that this solves?  What's the use case and how does it help?  If this were not merged, would DMD be in any kind of deficit?

**Comment by marler8997** [comment]
My use case was debugging the contents of symbol tables.

File: src/dmd/dsymbol.d [hunk 2]
```diff
@@ -2114,6 +2114,22 @@ extern (C++) final class DsymbolTable : RootObject
         return s;
     }
 
+    debug
+    {
+        /**
+        dump the symbol table contents
+        */
+        void dump()
```
Comment (JinShil) [inline src/dmd/dsymbol.d]: This could probably be renamed to something like "printSymbolTable".

Comment (marler8997) [reply src/dmd/dsymbol.d]: seems redundant...
```
symbolTable.printSymbolTable
```

What's the argument for the redundancy?

Comment (JinShil) [reply src/dmd/dsymbol.d]: Sorry. I didn't realize it was a member.  Still "dump" is quite informal (slang in a way); perhaps "print"

Comment (marler8997) [reply src/dmd/dsymbol.d]: ok

**Comment by JinShil** [comment]
Is there any way to test this in the test suite to keep it from getting accidentally breaking, and provide a way to verify the output without having to test it locally?

**Comment by marler8997** [comment]
The logic that would need to be tested is AARange. Would a unnittest suffice?

**Comment by marler8997** [comment]
> I'd also like to verify the output. Here's and idea, but I don't know if it will work.

I'm all for testing...but, I do think you can "overtest" as well.  The `print` function itself is just a `foreach` loop with a print statement.  All it's doing is printing the contents of the associative array.

That being said, the underlying logic in `AARange` would definitely warrant a unittest to prevent regression.

This is of course my opinion, maybe I'm wrong and this is worth extra testing.  If it helps the likelihood that it will be accepted than I'm willing to accommodate, however, I have a feeling that it actually might hurt the odds rather than help them.

File: src/dmd/root/aav.d [hunk 3]
```diff
@@ -121,6 +127,63 @@ extern (C++) Value dmd_aaGetRvalue(AA* aa, Key key)
...
+    */
+    @property auto asRange(AA* aa)
+    {
+        return AARange(aa);
+    }
+
+    private struct AARange
+    {
+        AA* aa;
+        size_t b_index;
```
Comment (wilzbach) [inline src/dmd/root/aav.d]: DStyle: `dIndex` (and a comment wouldn't hurt too)

File: src/dmd/root/aav.d [hunk 4]
```diff
@@ -121,6 +127,63 @@ extern (C++) Value dmd_aaGetRvalue(AA* aa, Key key)
     return null; // not found
 }
 
+debug
+{
+    /**
+    Returns a range of key/values for `aa`.
+    Returns: a range of key/values for `aa`.
```
Comment (wilzbach) [inline src/dmd/root/aav.d]: duplicate?

**Comment by wilzbach** [comment]
> Decided to make a PR for it if other people think it's useful to include.

Yes, it would be quite useful. I have missed this many times already. Thanks a lot for submitting this.

> The logic that would need to be tested is AARange. Would a unnittest suffice?

I think so too. It's not like it would be a big deal if an opt-in debug function stops to work and I don't think we have a precedent of testing `debug` code in DMD except for using D's built-in unittest blocks.
Also it looks like rather big modifications to the Makefile are required to be able to test this in the testsuite which I think is an overkill.

File: src/dmd/root/aav.d [hunk 5]
```diff
@@ -121,6 +127,63 @@ extern (C++) Value dmd_aaGetRvalue(AA* aa, Key key)
...
+                toNext();
+            }
+        }
+
+        private void toNext()
+        {
+            for (; b_index < aa.b_length; b_index++)
+            {
+                auto next = aa.b[b_index];
+                if (next)
```
Comment (wilzbach) [inline src/dmd/root/aav.d]: One line: `if (auto next = ...)`

File: src/dmd/root/aav.d (hunk 4 above)
Comment (marler8997) [reply src/dmd/root/aav.d]: Actually it was intentional, I was trying to make a valid ddoc comment.  A description and a return value...maybe one of them is not necessary?

File: src/dmd/root/aav.d (hunk 5 above)
Comment (marler8997) [reply src/dmd/root/aav.d]: actually that logic doesn't work...that assignment can't be done until after the loop condition is satisfied.

File: src/dmd/dsymbol.d (hunk 2 above)
Comment (marler8997) [reply src/dmd/dsymbol.d]: uh...it looks like the `print` function is already taken...switched it back to `dump` unless you have another name you'd prefer.

Comment (JinShil) [reply src/dmd/dsymbol.d]: What does `print()` currently do?  Maybe it should be overridden.

Comment (marler8997) [reply src/dmd/dsymbol.d]: ok, fine with me :)

**Comment by jacob-carlborg** [comment]
I don't mind some extra debug functionality in DMD. I think it's very useful, especially for someone not that familiar with the code base.

File: src/dmd/root/aav.d [hunk 6]
```diff
@@ -121,6 +127,93 @@ extern (C++) Value dmd_aaGetRvalue(AA* aa, Key key)
...
+                toNext();
+            }
+        }
+
+        private void toNext()
+        {
+            for (; bIndex < aa.b_length; bIndex++)
+            {
+                auto next = aa.b[bIndex];
+                if (next)
```
Comment (wilzbach) [inline src/dmd/root/aav.d]: > One line: if (auto next = ...)
> actually that logic doesn't work...that assignment can't be done until after the loop condition is satisfied.

I don't follow you here - is this a remark on previous code?

Comment (marler8997) [reply src/dmd/root/aav.d]: Woops, I thought you were talking about putting `auto next = ...` in the for loop...will change

**Review by wilzbach — APPROVED** [review]

**Comment by wilzbach** [comment]
BTW I'm wondering whether it would make sense to start abstracting over the relicts from the C/C++ age (like it's done for arrays) and use D-syntax for associative arrays, e.g. sth. like this:

```d
struct DAa(K, V)
{
    AA* aa;
    size_t length()
    {
        return dmd_aaLen(aa);
    }

    V* opIndex(K key)
    {
        return cast(V*) dmd_aaGet(&aa, cast(void*) key);
    }

    auto opIndexAssign(V value, K key)
    {
        auto ptr = dmd_aaGet(&aa, cast(void*) key);
        *ptr = cast(void*) value;
        return *ptr;
    }
}
```

**Comment by marler8997** [comment]
Yeah I think that's a great idea.

File: src/dmd/root/aav.d [hunk 7]
```diff
@@ -121,6 +127,92 @@ extern (C++) Value dmd_aaGetRvalue(AA* aa, Key key)
     return null; // not found
 }
 
+debug
+{
+    /**
+    Returns a range of key/values for `aa`.
+
+    Returns: a range of key/values for `aa`.
```
Comment (JinShil) [inline src/dmd/root/aav.d]: Nit:  Could use "Gets a range..." for the summary to avoid the redundancy.

**Review by JinShil — APPROVED** [review]

**Review by wilzbach — APPROVED** [review]

**Comment by timotheecour** [comment]
@marler8997 

Thanks for dong this PR, something I also wished existed; however, one thing i don't like is:
`override void print()` 
how about instead having:
`char* toString()` or `extern(D) string toString()` or another name like `toStringDebug`.

forcing print to stdout is not flexible, eg:
* in my case I want to log through a function that prepends  __FILE__:__LINE__ before calling toString; * another case is recursive dumping using indentation that depends on nesting level
* another case is dumping to a file insetead of stdout

also, it's much more standard to output to a string/char* than to dump to stdout

**Pull request merged**